linksharing/handler.go

```diff
@@ -307,3 +332,103 @@ func makeLocation(base *url.URL, reqPath string) string {
 	location.Path = path.Join(location.Path, reqPath)
 	return location.String()
 }
+
+func (handler *Handler) handleHostingService(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	serializedAccess, root, err := handler.getRootAndAccess(r.Host)
+	if err != nil {
+		handler.log.Error("unable to handle request", zap.Error(err))
+		http.Error(w, "unable to handle request", http.StatusInternalServerError)
+		return err
+	}
+	access, err := uplink.ParseAccess(serializedAccess)
+	if err != nil {
+		handler.handleUplinkErr(w, "parse access", err)
+		return err
+	}
+	project, err := uplink.OpenProject(ctx, access)
+	if err != nil {
+		handler.handleUplinkErr(w, "open project", err)
+		return err
+	}
+	defer func() {
+		if err := project.Close(); err != nil {
+			handler.log.With(zap.Error(err)).Warn("unable to close project")
+		}
+	}()
+
+	// clean path and reconstruct
+	rootPath := strings.SplitN(root, "/", 2)
+	bucket := rootPath[0]
+	path := strings.TrimPrefix(r.URL.Path, "/")
+	path = strings.TrimSuffix(path, "/")
+	if len(rootPath) == 2 {
+		pathPrefix := strings.TrimPrefix(rootPath[1], "/")
+		pathPrefix = strings.TrimSuffix(pathPrefix, "/")
+		path = pathPrefix + "/" + path
+	}
+
+	if path == "" {
+		path = "index.html"
+	}
+
+	o, err := project.StatObject(ctx, bucket, path)
+	if err != nil {
+		handler.handleUplinkErr(w, "stat object", err)
+		return err
+	}
+
+	httpranger.ServeContent(ctx, w, r, path, o.System.Created, newObjectRanger(project, o, bucket))
+	return nil
+}
+
+func (handler *Handler) getRootAndAccess(hostname string) (serializedAccess, root string, err error) {
+	handler.txtRecords.mu.Lock()
+	defer handler.txtRecords.mu.Unlock()
+	//check cache for access and root
+	record, ok := handler.txtRecords.cache[hostname]
+	// do a txt record lookup if the cache doesn't contain a corresponding entry or if the entry is expired
+	if !ok || record.timestamp.Add(handler.txtRecords.ttl).Before(time.Now()) {
+		records, err := net.LookupTXT(hostname)
+		if err != nil {
+			return serializedAccess, root, err
+		}
+		serializedAccess, root, err = parseRecords(records)
+
+		if err != nil {
+			return serializedAccess, root, err
+		}
+	}
+	// update cache
+	handler.txtRecords.cache[hostname] = txtRecord{access: serializedAccess, root: root, timestamp: time.Now()}
+
+	return serializedAccess, root, err
+}
+
+func parseRecords(records []string) (serializedAccess, root string, err error) {
```

Comment:
An example of a record and what this is looking for would be helpful.

Reply:
done